pym/_emerge/depgraph.py

```diff
@@ -3300,13 +3300,13 @@ def _add_pkg_deps(self, pkg, allow_unsatisfied=False):
 		if removal_action:
 			depend_root = myroot
 		else:
-			if eapi_attrs.bdepend or eapi_attrs.hdepend:
+			if eapi_attrs.hdepend:
 				depend_root = myroot
 			else:
 				depend_root = self._frozen_config._running_root.root
 				root_deps = self._frozen_config.myopts.get("--root-deps")
 				if root_deps is not None:
-					if root_deps is True:
+					if root_deps is True or eapi_attrs.bdepend:
 						depend_root = myroot
```

Comment:
My hope was the the `--root-deps` option would become obsolete, since it was intended only as a workaround in the absence of native cross-compilation support in the EAPI. This is why `eapi_attrs.hdepend` behavior was independent of the `--root-deps` option. If `eapi_attrs.bdepend` needs to interact with the `--root-deps` option, then I'd like some kind of explanation.

Reply:
I explained what the option would do against this EAPI in 0fc532a4524b48a62f2178fa7a2298312a4e81a4. As for why, we're allowing users to choose between building against headers/libraries in / and ROOT. Both are valid use cases although only the latter is an option for cross-compiling. 5-hdepend forgot to consider the bigger picture and this would have annoyed plenty of users.

I initially didn't think we'd use the --root-deps option for this and expected to just take the value of SYSROOT alone. Given that --root-deps already exists, this might be a confusing inconsistency. crossdev already uses --root-deps=rdeps by default, which is useful if we ignore the rdeps part for EAPI 7.

Reply:
I see the commit message mentions something about the `ROOT!=/ SYSROOT=/ use case`, so I'd need some clarification on why that introduces a need for `--root-deps`. My expectation was that we would be using ebuild variables to specify where dependencies are installed, rather than emerge options.

Reply:
For the `ROOT!=/ SYSROOT=/`  use case, you wouldn't give --root-deps. The problem is that 5-hdepend effectively enabled --root-deps all the time and that implied SYSROOT=ROOT. There was no way to prevent that so 7 gives you back that control.

Reply:
Based on your explanation, I think that for eapi_attrs.bdepend, depend_root should be purely a function of the SYSROOT (and independent of the `--root-deps` option, as I have explained earlier).

Reply:
I'm okay with that but if `SYSROOT` is set to some other location then I think we'll just have to bail out instead of trying to accommodate existing setups.

crossdev's `cross-emerge` currently sets `SYSROOT=${EPREFIX}/usr/${CHOST}`, which is the same as the default `ROOT` value but if you change `ROOT` to something else then `SYSROOT` will not change accordingly. Perhaps this could be fixed.

You probably have the best idea of what it would take to actually support a third location for `SYSROOT`. Is it as much work as I suspect it would be? It would share `PORTAGE_CONFIGROOT` with `ROOT` so it wouldn't be an entirely separate configuration.

Reply:
We can have backward-compatible SYSROOT pass-through as the default behavior, and add a `--sysroot=SYSROOT` emerge option that explicitly enables new behavior for EAPI 7.

Reply:
Let's document each of the configurations that can exist simultaneously. Prior to EAPI 7, there were a maximum of 2 configurations existing simultaneously, each having independent sets of variables:

running root: ROOT=/ + PORTAGE_CONFIGROOT=/
target root: ROOT=${FOO} + PORTAGE_CONFIGROOT=${BAR}

For EAPI 7, the configuration for the "running root" should not export SYSROOT to the ebuild environment, since its purpose is to build packages for `ROOT=/` in the classic manner. The configuration for the "target root" should export SYSROOT to the ebuild environment. For EAPI 7, we can introduce a third configuration with `ROOT=${SYSROOT}` and `PORTAGE_CONFIGROOT=${BAZ}` (maybe add a `--sysroot-configroot` option if necessary).

Reply:
I hadn't even considered `ROOT=/ SYSROOT!=/` but I think you're right, that would be madness so let's avoid it.

On your last point, my brain is currently doing backflips so I can't work out if you're wrong, I'm wrong, or I've just misunderstood what you said. Last night's visit to the hospital at 3am probably isn't helping. I'll come back after more sleep.

Reply:
The gist of it is, we've got at least 2 separate configurations to deal with, each with mostly independent environment variables. We need to carefully define which variables are available to each configuration. For example, `pym/portage/__init__.py` we have a clean_env dictionary which contains variables that are are shared by all of the configurations. It may be that only one of the configurations really needs a SYSROOT variable where SYSROOT != ROOT, and another one might set `ROOT=${SYSROOT}`.

Reply:
Sleep has helped. I think that even under EAPI 7, this boils down to 2 situations when `ROOT!=/`.

## SYSROOT=/
In this situation, the same PORTAGE_CONFIGROOT must be applied to BDEPEND (/), DEPEND (/) and RDEPEND (ROOT). Why? BDEPEND and DEPEND must have matching configurations if they're both installing into the same location. RDEPEND and DEPEND must also have matching configurations because you can't, for instance, build against OpenSSL, and then run against LibreSSL. You also have to consider existing packages on / that may conflict.

```
$ emerge -pv socat
[ebuild   R    ] net-misc/socat-1.7.3.2::gentoo  USE="ipv6 readline ssl -bindist -libressl -tcpd" 0 KiB

$ ROOT=/tmp/foo USE=libressl emerge -pv socat
[ebuild  N     ] app-misc/c_rehash-1.7-r1::gentoo to /tmp/foo/ 0 KiB
[ebuild  N     ] sys-apps/debianutils-4.8.4::gentoo to /tmp/foo/ USE="installkernel -static" 0 KiB
[ebuild  N     ] dev-libs/libressl-2.6.4:0/44::gentoo  USE="asm -static-libs" ABI_X86="32 (64) (-x32)" 3,153 KiB
[ebuild  N     ] dev-libs/libressl-2.6.4:0/44::gentoo to /tmp/foo/ USE="asm -static-libs" ABI_X86="32 (64) (-x32)" 0 KiB
[ebuild  N     ] app-misc/ca-certificates-20170717.3.36::gentoo to /tmp/foo/ USE="-cacert -insecure_certs" 22,486 KiB
[ebuild  N     ] sys-libs/ncurses-6.1-r2:0/6::gentoo to /tmp/foo/ USE="cxx gpm threads unicode -ada -debug -doc -minimal -profile -static-libs {-test} -tinfo -trace" ABI_X86="32 (64) (-x32)" 3,287 KiB
[ebuild  N     ] sys-libs/gpm-1.20.7-r2::gentoo to /tmp/foo/ USE="(-selinux) -static-libs" ABI_X86="32 (64) (-x32)" 0 KiB
[ebuild  N     ] sys-libs/readline-7.0_p3:0/7::gentoo to /tmp/foo/ USE="-static-libs -utils" ABI_X86="32 (64) (-x32)" 0 KiB
[ebuild  N     ] net-misc/socat-1.7.3.2::gentoo to /tmp/foo/ USE="ipv6 libressl readline ssl -bindist -tcpd" 0 KiB
[blocks B      ] dev-libs/openssl:0 ("dev-libs/openssl:0" is blocking dev-libs/libressl-2.6.4)
```

These are just some of the many things that could go wrong with differing configurations. You can already fall into this trap right now, it's not a new problem arising from EAPI 7.

## SYSROOT!=/
Here, the same PORTAGE_CONFIGROOT must be applied to DEPEND (SYSROOT) and RDEPEND (ROOT) for the same reason given above but BDEPEND (/) may be different. Because we only have one PORTAGE_CONFIGROOT variable, we assume the configuration for BDEPEND to be in /.

## Conclusions
Thinking about all this has made me realise a few things.

Your suggestion of `--sysroot-configroot` doesn't fit anywhere. In fact, the whole notion of a hypothetical third location for SYSROOT doesn't seem all that useful because with the introduction of BDEPEND, DEPEND will equal RDEPEND in most cases anyway. If we did support it, you certainly wouldn't want to build these packages twice. Perhaps Portage could automatically install binary packages created from SYSROOT into ROOT but what would be the point? Sure, it would avoid the odd static library or header from being put into ROOT but embedded system builders can trivially remove these, along with all the other headers that will inevitably end up in ROOT.

I noticed that PORTAGE_CONFIGROOT is effective even when `ROOT=/`. This is confusing because as soon as you set ROOT to be something else, it applies to that location instead of /. Sure, this is true for most environment variables like USE and CFLAGS but unlike those, you obviously can't set PORTAGE_CONFIGROOT in make.conf.

When I consider that, plus the restrictions on PORTAGE_CONFIGROOT that I have outlined above, I start to wonder whether it would be simpler to just do away with PORTAGE_CONFIGROOT entirely and use SYSROOT for this purpose instead. In all cases, the configuration needs to follow SYSROOT. The ability to put the configuration in some arbitrary location doesn't seem that useful and just complicated matters for both us and the users. You might not want /etc/portage in your final system but just like headers, this can trivially be removed.

What do you think?

Reply:
> RDEPEND (ROOT) and DEPEND (/)  must also have matching configurations because you can't, for instance, build against OpenSSL, and then run against LibreSSL.

In order to keep things simple, that's a fair assumption to make for now. However, they only need to be compatible, not necessarily identical. Our code should try to be flexible enough to allow some deviation in the configurations, though we don't necessarily have to expose this kind of flexibility to emerge users.

> Because we only have one PORTAGE_CONFIGROOT variable, we assume the configuration for BDEPEND to be in /.

This is also a user interface decision. It's fine to do it that way for now, but internally our code should try to be flexible enough to allow for deviation from that.

> I start to wonder whether it would be simpler to just do away with PORTAGE_CONFIGROOT entirely and use SYSROOT for this purpose instead.

That's fine, but in order to maintain backward compatibility I'd like to hide any breaking changes behind commandline options. For example, we can make usage of a new --sysroot=SYSROOT option be incompatible with --config-root=PORTAGE_CONFIGROOT.

Reply:
Thanks, that helps a lot. I may still need a hand with the code as you're hard to please but I'll give it my best shot. :wink: